File: mloggerai-py/mloggerai/errorsolver.py
import logging
import os
from logging.handlers import RotatingFileHandler
from pathlib import Path
from typing import cast

from dotenv import load_dotenv
from openai import OpenAI
from openai.types.chat import ChatCompletionMessageParam


class ErrorSolver:
    def __init__(
            self,
            model=None,
            base_url=None,
            api_key=None,
            log_file="logs/logger.log",
            log_level=logging.DEBUG,
            output_language="italiano",
            temperature=0.3
    ):

        load_dotenv()  # Carica le variabili da .env
        self.base_url: str = os.getenv("OPENAI_API_URL", "http://localhost:1234/v1")
        self.api_key: str = os.getenv("OPENAI_API_KEY", "")
        self.model: str = os.getenv("OPENAI_API_MODEL", None)
        self.model: str = os.getenv("OPENAI_API_MODEL", None)
        self.temperature = temperature
        if model is not None:
            self.model = model
        if base_url is not None:
            self.base_url = base_url
        if api_key is not None:
            self.api_key = api_key

        self.client = OpenAI(base_url=self.base_url, api_key=self.api_key)

        # Logger principale
        self.logger = logging.getLogger("AppLogger")
        self.logger.setLevel(log_level)
        self.output_language = output_language
        if not self.logger.hasHandlers():
            formatter = logging.Formatter("%(asctime)s - %(levelname)s - %(message)s")

            # Console handler
            ch = logging.StreamHandler()
            ch.setFormatter(formatter)
            self.logger.addHandler(ch)

            # File handler
            Path(log_file).parent.mkdir(parents=True, exist_ok=True)
            fh = RotatingFileHandler(log_file, maxBytes=5_000_000, backupCount=3)
            fh.setFormatter(formatter)
            self.logger.addHandler(fh)

        # Aggancia handler AI
        self._attach_ai_handler(ai_level=logging.ERROR)

    def solve_from_log(self, text):
        try:
            system_prompt: str = os.getenv(
                "OPENAI_API_PROMPT",
                f"Trova il bug e proponi la soluzione in modo molto conciso.",
            )
            messages = cast(
                list[ChatCompletionMessageParam],
                [
                    {"role": "system", "content": f"{system_prompt}\nRispondi sempre in lingua {self.output_language} presentendo un solo esempio di codice non errato"},
                    {"role": "user", "content": text},
                ],
            )
            completion = self.client.chat.completions.create(
                model=self.model, temperature=self.temperature, max_tokens=180, messages=messages
            )
            return completion.choices[0].message.content.strip()
        except Exception as e:
            return f"Errore AI: {e}"

    def _attach_ai_handler(self, ai_level=logging.ERROR):
        solver = self

        class AIHandler(logging.Handler):
            def emit(self, record):
                if getattr(record, "_from_ai_solver", False):
                    return
                try:
                    msg = self.format(record)
                    solution = solver.solve_from_log(msg)
                    combined = f"📘 Soluzione AI: {solution}"
                    solver.logger.debug(combined, extra={"_from_ai_solver": True})
                except Exception as e:
                    solver.logger.debug(
                        f"Errore AI interno: {e}", extra={"_from_ai_solver": True}
                    )

        handler = AIHandler()
        handler.setLevel(ai_level)
        handler.setFormatter(
            logging.Formatter("%(asctime)s - %(levelname)s - %(message)s")
        )
        self.logger.addHandler(handler)

File: mloggerai-ts/src/ErrorSolver.ts
import fs from "fs";
import path from "path";
import dotenv from "dotenv";
import { fileURLToPath } from "url";
import winston, { Logger } from "winston";
import { OpenAI } from "openai";

dotenv.config();

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export interface ErrorSolverConfig {
  model?: string | null;
  base_url?: string | null;
  api_Key?: string | null;
  temperature?: number;
  max_tokens?: number;
  base_prompt?: string | null;
  logFile?: string;
  logLevel?: string;
  outputLanguage?: string;
}

export class ErrorSolver {
  private baseURL: string;
  private apiKey: string;
  private model: string;
  private client: OpenAI;
  private outputLanguage: string;
  private temperature: number;
  private max_tokens: number;
  private systemPrompt: string;
  private logger: Logger;

  constructor({
    model = null,
    base_url = null,
    api_Key = null,
    temperature = 0.3,
    max_tokens = 150,
    base_prompt = null,
    logFile = "logs/logger.log",
    logLevel = "debug",
    outputLanguage = "italiano",
  }: ErrorSolverConfig = {}) {
    this.baseURL =
      base_url || process.env.OPENAI_API_URL || "http://localhost:1234/v1";
    this.apiKey = api_Key || process.env.OPENAI_API_KEY || "";
    this.model = model || process.env.OPENAI_API_MODEL || "";
    this.client = new OpenAI({
      baseURL: this.baseURL,
      apiKey: this.apiKey,
    });

    this.outputLanguage = outputLanguage;
    this.temperature = temperature;
    this.max_tokens = max_tokens;
    this.systemPrompt =
      base_prompt ||
      process.env.OPENAI_API_PROMPT ||
      "Trova il bug e proponi la soluzione in modo molto conciso fornendo anche un solo esempio di codice corretto";
    // Ensure log directory exists without constructing fs.Stats manually
    const logDir = path.dirname(logFile);
    try {
      if (!fs.existsSync(logDir)) {
        fs.mkdirSync(logDir, { recursive: true });
      }
    } catch (err) {
      console.error(`Cannot create log directory: ${err}`);
    }

    this.logger = winston.createLogger({
      level: logLevel,
      format: winston.format.combine(
        winston.format.timestamp(),
        winston.format.printf(
          ({ timestamp, level, message }) =>
            `${timestamp} - ${level.toUpperCase()} - ${message}`
        )
      ),
      transports: [
        new winston.transports.Console(),
        new winston.transports.File({
          filename: logFile,
          maxsize: 5_000_000,
          maxFiles: 3,
        }),
      ],
    });

    this._attachAIHandler("error");
  }

  async solveFromLog(text: string): Promise<string> {
    try {
      // 🔹 Log info delle variabili principali
      this.logger.debug(`baseURL: ${this.baseURL}`);
      this.logger.debug(`apiKey: ${this.apiKey ? "[SET]" : "[EMPTY]"}`); // nascondi chiave completa
      this.logger.debug(`model: ${this.model}`);
      this.logger.debug(`outputLanguage: ${this.outputLanguage}`);
      this.logger.debug(`temperature: ${this.temperature}`);
      this.logger.debug(`max_tokens: ${this.max_tokens}`);
      this.logger.debug(`systemPrompt: ${this.systemPrompt}`);

      const completion = await this.client.chat.completions.create({
        model: this.model,
        temperature: this.temperature,
        max_tokens: this.max_tokens,
        messages: [
          {
            role: "system",
            content: `${this.systemPrompt}. Rispondi sempre in lingua ${this.outputLanguage}`,
          },
          { role: "user", content: text },
        ],
      });

      return completion.choices[0].message?.content?.trim() || "";
    } catch (err: any) {
      return `Errore AI: ${err.message}`;
    }
  }

  private _attachAIHandler(level: keyof Logger): void {
    const origLog = (this.logger[level] as Function).bind(this.logger);
    (this.logger[level] as any) = async (message: string, ...args: unknown[]) => {
      origLog(message, ...args);
      try {
        const solution = await this.solveFromLog(message);
        this.logger.debug(`🧠💡 Soluzione AI: ${solution}`);
        return solution;
      } catch (err: any) {
        this.logger.debug(`Errore AI interno: ${err.message}`);
        return undefined;
      }
    };
  }
}
